dotnet-authserver/src/TeacherIdentity.AuthServer/Helpers/QueryStringSignatureHelper.cs

```diff
@@ -0,0 +1,49 @@
+using System.Security.Cryptography;
+using System.Text;
+using Flurl;
+
+namespace TeacherIdentity.AuthServer.Helpers;
+
+public class QueryStringSignatureHelper
+{
+    private const string SignatureParameterName = "sig";
+
+    private readonly byte[] _keyBytes;
+
+    public QueryStringSignatureHelper(string key)
+    {
+        _keyBytes = Encoding.UTF8.GetBytes(key);
+    }
+
+    public Url AppendSignature(Url url, string[] parameterNames)
+    {
+        var sig = CalculateSignature(url, parameterNames);
+        return url.SetQueryParam(SignatureParameterName, sig);
+    }
+
+    public bool VerifySignature(Url url, string[] parameterNames)
+    {
+        var sig = CalculateSignature(url, parameterNames);
+        return sig == url.QueryParams.SingleOrDefault(qp => qp.Name == SignatureParameterName).Value?.ToString();
+    }
+
+    private string CalculateSignature(Url url, string[] parameterNames)
+    {
+        var canonicalizedValues =
```

Comment:
Can we not just hash the whole thing?

Reply:
We can! I thought it might cause problems in a few places but all the tests are still green so I'll go ahead with that.